File: frontend/src/types/worldTick.ts
import type { ActionId, WorldData } from './world'

export type TargetKind = 'location' | 'npc'

export interface WorldActionInfo {
  id: number
  clock_tick: number
  actor_id: string
  action_type: ActionId
  target_kind: TargetKind | null
  target_id: string | null
  reason: string
  status: 'executed'
  run_id: string
  proposal_id: number
  world_version: number
  world_time: string
}

export interface WorldEventInfo {
  id: number
  run_id: string | null
  world_version: number
  clock_tick: number
  event_sequence: number
  event_type: string
  actor_id: string | null
  action_id: number | null
  source_event_id: number | null
  description: string
  world_time: string
  payload: Record<string, unknown>
  visibility: string
  secrecy: string
  causation_id: string | null
  correlation_id: string
  created_at: string
}

export interface AgentRunSummary {
  id: string
  mode: 'deterministic'
  trigger_type: 'world_advance'
  status: 'completed'
  base_world_version: number
  resulting_world_version: number
  base_clock_tick: number
  resulting_clock_tick: number
}

export interface WorldTickData {
  run: AgentRunSummary
  world: WorldData
  actions: WorldActionInfo[]
  events: WorldEventInfo[]
}

File: backend/app/schemas/world_tick.py
from typing import Literal

from pydantic import BaseModel, Field

from backend.app.schemas.world import ActionId, WorldData


class WorldTickRequest(BaseModel):
    expected_tick: int = Field(ge=0)


class WorldActionInfo(BaseModel):
    id: int
    tick: int = Field(ge=1)
    actor_id: str
    action_type: ActionId
    target_kind: Literal["location", "npc"] | None
    target_id: str | None
    reason: str
    status: Literal["recorded"]
    world_time: str


class WorldEventInfo(BaseModel):
    id: int
    tick: int = Field(ge=1)
    event_type: Literal["npc_action"]
    actor_id: str
    action_id: int
    description: str
    world_time: str


class WorldTickData(BaseModel):
    world: WorldData
    actions: list[WorldActionInfo]
    events: list[WorldEventInfo]
